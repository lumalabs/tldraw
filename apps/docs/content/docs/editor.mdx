---
title: Styles
status: published
author: steveruizok
date: 3/22/2023
order: 1
keywords:
  - ui
  - app
  - editor
  - control
  - select
---

# Markdown Stylesheet

## Headers

# H1 - Main Title

## H2 - Section Title

### H3 - Subsection Title

#### H4 - Smaller Section

##### H5 - Minor Section

###### H6 - Smallest Header

## Paragraphs

Regular paragraph text with **bold**, _italic_, and `inline code` formatting.

## Blockquotes

> This is a standard blockquote.
>
> > Nested blockquote for additional emphasis.

## Lists

### Unordered List

- Item 1
- Item 2
  - Subitem 2.1
  - Subitem 2.2
    - Sub-subitem 2.2.1
- Item 3

### Ordered List

1. First item
2. Second item
   1. Subitem 2.1
   2. Subitem 2.2
3. Third item

### Task List

- [x] Completed task
- [ ] Pending task
- [ ] Another pending task

## Code Blocks

### Inline Code

Example: `print("Hello, World!")`

### Block Code

```python
# Python example
def greet(name):
    return f"Hello, {name}!"
print(greet("Alice"))
```

```js
// JavaScript example
function greet(name) {
	return `Hello, ${name}!`
}
console.log(greet('Alice'))
```

## Tables

| Column 1 | Column 2 | Column 3  |
| -------- | -------- | --------- |
| Row 1    | Data     | More Data |
| Row 2    | Data     | More Data |
| Row 3    | Data     | More Data |

## Horizontal Rule

---

## Links and Images

[Example Link](https://example.com)

![Example Image](https://via.placeholder.com/150)

## Emphasis

**Bold Text**  
_Italic Text_  
~~Strikethrough~~

## Footnotes

Here is a sentence with a footnote.[^1]

[^1]: This is the footnote explanation.

## Definition Lists

Term 1  
: Definition 1

Term 2  
: Definition 2

## Custom HTML (If Supported)

<details>
	<summary>Click to expand</summary>
	Hidden content here.
</details>

## Escaping Characters

Use a backslash to escape special characters: \*not italicized\*
