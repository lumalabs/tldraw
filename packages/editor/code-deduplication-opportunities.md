# Code Deduplication Opportunities in Editor Package

This document outlines opportunities to remove repeated code patterns in the `packages/editor` package, organized by category and priority.

## 🔧 High Priority - Event Listener Management Patterns

### Pattern: useEffect + addEventListener + removeEventListener
**Found in**: Multiple hooks (`useDocumentEvents`, `usePassThroughWheelEvents`, `usePassThroughMouseOverEvents`, `useScreenBounds`, `useViewportHeight`, `useCoarsePointer`, etc.)

**Current Duplication**:
```typescript
useEffect(() => {
    const handleEvent = (e: Event) => { /* logic */ }
    element.addEventListener('eventType', handleEvent, options)
    return () => {
        element.removeEventListener('eventType', handleEvent)
    }
}, deps)
```

**Refactoring Opportunity**:
Create a custom hook `useEventListener(element, eventType, handler, options, deps)` to consolidate this pattern.

**Files Affected**: 
- `useDocumentEvents.ts` (multiple event listeners)
- `usePassThroughWheelEvents.ts`
- `usePassThroughMouseOverEvents.ts` 
- `useScreenBounds.ts`
- `useViewportHeight.ts`
- `useCoarsePointer.ts`
- `useFixSafariDoubleTapZoomPencilEvents.ts`

## 🎯 Medium Priority - Pass-Through Event Patterns

### Pattern: Event Redispatching to Canvas
**Found in**: `usePassThroughWheelEvents`, `usePassThroughMouseOverEvents`, and `useDocumentEvents`

**Current Duplication**:
```typescript
function handleEvent(e: Event) {
    if ((e as any).isSpecialRedispatchedEvent) return
    preventDefault(e)
    const cvs = container.querySelector('.tl-canvas')
    if (!cvs) return
    const newEvent = new EventType(e.type, e as any)
    ;(newEvent as any).isSpecialRedispatchedEvent = true
    cvs.dispatchEvent(newEvent)
}
```

**Refactoring Opportunity**:
Create a utility function `redispatchEventToCanvas(container, originalEvent, EventConstructor)` to consolidate this logic.

**Files Affected**:
- `usePassThroughWheelEvents.ts` 
- `usePassThroughMouseOverEvents.ts`
- `useDocumentEvents.ts` (drag events)

## 🏗️ Medium Priority - Shape Bounds Calculation Patterns

### Pattern: getShapePageBounds Usage
**Found in**: Multiple files with similar bounds checking patterns

**Current Duplication**:
```typescript
const bounds = editor.getShapePageBounds(shape)
if (!bounds) return // or similar null checks
// Use bounds for calculations
```

**Refactoring Opportunity**:
Create utility functions for common bounds operations:
- `getRequiredShapePageBounds(editor, shape)` - throws if bounds null
- `getShapePageBoundsOrDefault(editor, shape, defaultBounds)`
- `ensureShapePageBounds(editor, shapes)` - filters out shapes without bounds

**Files Affected**:
- `Editor.ts` (25+ occurrences)
- `SnapManager.ts` and related snap files
- `DefaultCanvas.tsx`
- `reorderShapes.ts`

## 🔄 Low Priority - Editor State Access Patterns

### Pattern: this.editor Access in Managers
**Found in**: Multiple manager classes

**Current Duplication**:
```typescript
this.editor.getContainer()
this.editor.getInstanceState()
this.editor.getSelectedShapes()
// etc.
```

**Refactoring Opportunity**:
Create base manager class with common editor access methods or use dependency injection for commonly accessed editor methods.

**Files Affected**:
- `FocusManager.ts`
- `SnapManager.ts`
- `FontManager.ts`
- `TextManager.ts`
- `ScribbleManager.ts`

## 🎪 Low Priority - Hook Parameter Validation

### Pattern: Ref Validation in Hooks
**Found in**: Multiple custom hooks

**Current Duplication**:
```typescript
if (!ref) throw Error('useHookName must be passed a ref')
```

**Refactoring Opportunity**:
Create a utility function `validateHookRef(ref, hookName)` or use a custom hook wrapper.

**Files Affected**:
- `usePassThroughWheelEvents.ts`
- `usePassThroughMouseOverEvents.ts`

## 📝 Utility Function Opportunities

### Pattern: Event Prevention
**Found in**: `dom.ts` exports `preventDefault` and `stopEventPropagation` but could be enhanced

**Refactoring Opportunity**:
Create combined utility functions:
- `preventAndStopEvent(e)` - combines preventDefault + stopPropagation
- `conditionalPreventDefault(e, condition)` - prevents based on condition

### Pattern: MediaQuery Management
**Found in**: `useDocumentEvents.ts` has complex MediaQuery handling that could be extracted

**Refactoring Opportunity**:
Create `useMediaQuery(query, callback)` hook to handle addEventListener/removeEventListener for media queries.

## 🚀 Implementation Strategy

### Phase 1: High-Impact, Low-Risk
1. Create `useEventListener` hook
2. Create event redispatch utilities
3. Replace 3-5 most obvious duplicated patterns

### Phase 2: Medium-Impact Refactoring  
1. Create shape bounds utility functions
2. Implement base manager class patterns
3. Create media query management hook

### Phase 3: Polish and Optimization
1. Add parameter validation utilities
2. Create combined event utilities
3. Review and consolidate any remaining patterns

## 📊 Impact Assessment

**Lines of Code Reduction**: Estimated 200-300 lines
**Maintenance Improvement**: Significant - centralized event handling patterns
**Testing Benefits**: Easier to test common patterns in isolation
**Performance Impact**: Minimal to none, possibly slight improvement from shared functions

## 🛠️ Recommended Next Steps

1. **Start with `useEventListener` hook** - highest impact, safest change
2. **Create event redispatch utilities** - clear pattern, easy to test
3. **Gradually migrate existing hooks** - one at a time to minimize risk
4. **Add comprehensive tests** for new utility functions
5. **Update documentation** to encourage use of new patterns

This analysis identified the most significant opportunities for code deduplication while maintaining the current functionality and improving maintainability.