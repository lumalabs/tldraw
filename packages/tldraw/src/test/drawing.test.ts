import { TLDrawShape, TLHighlightShape, last } from '@tldraw/editor'
import { TestEditor } from './TestEditor'
import { TEST_DRAW_SHAPE_SCREEN_POINTS } from './drawing.data'

jest.useFakeTimers()

let editor: TestEditor

afterEach(() => {
	editor?.dispose()
})

beforeEach(() => {
	editor = new TestEditor()

	editor.createShapes([])
})

type DrawableShape = TLDrawShape | TLHighlightShape

for (const toolType of ['draw', 'highlight'] as const) {
	describe(`When ${toolType}ing...`, () => {
		it('Creates a dot', () => {
			editor
				.setCurrentTool(toolType)
				.pointerDown(60, 60)
				.expectToBeIn(`${toolType}.drawing`)
				.pointerUp()
				.expectToBeIn(`${toolType}.idle`)

			expect(editor.getCurrentPageShapes()).toHaveLength(1)

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.type).toBe(toolType)
			expect(shape.props.segments.length).toBe(1)

			const segment = shape.props.segments[0]
			expect(segment.type).toBe('free')
		})

		it('Creates a dot when shift is held down', () => {
			editor
				.setCurrentTool(toolType)
				.keyDown('Shift')
				.pointerDown(60, 60)
				.expectToBeIn(`${toolType}.drawing`)
				.pointerUp()
				.expectToBeIn(`${toolType}.idle`)

			expect(editor.getCurrentPageShapes()).toHaveLength(1)

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.type).toBe(toolType)
			expect(shape.props.segments.length).toBe(1)

			const segment = shape.props.segments[0]
			expect(segment.type).toBe('straight')
		})

		it('Creates a free draw line when shift is not held', () => {
			editor.setCurrentTool(toolType).pointerDown(10, 10).pointerMove(20, 20)

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.props.segments.length).toBe(1)

			const segment = shape.props.segments[0]
			expect(segment.type).toBe('free')
		})

		it('Creates a straight line when shift is held', () => {
			editor.setCurrentTool(toolType).keyDown('Shift').pointerDown(10, 10).pointerMove(20, 20)

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.props.segments.length).toBe(1)

			const segment = shape.props.segments[0]
			expect(segment.type).toBe('straight')

			const points = segment.points
			expect(points.length).toBe(2)
		})

		it('Switches between segment types when shift is pressed / released  (starting with shift up)', () => {
			editor
				.setCurrentTool(toolType)
				.pointerDown(10, 10)
				.pointerMove(20, 20)
				.keyDown('Shift')
				.pointerMove(30, 30)
				.keyUp('Shift')
				.pointerMove(40, 40)
				.pointerUp()

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.props.segments.length).toBe(3)

			expect(shape.props.segments[0].type).toBe('free')
			expect(shape.props.segments[1].type).toBe('straight')
			expect(shape.props.segments[2].type).toBe('free')
		})

		it('Switches between segment types when shift is pressed / released (starting with shift down)', () => {
			editor
				.setCurrentTool(toolType)
				.keyDown('Shift')
				.pointerDown(10, 10)
				.pointerMove(20, 20)
				.keyUp('Shift')
				.pointerMove(30, 30)
				.keyDown('Shift')
				.pointerMove(40, 40)
				.pointerUp()

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.props.segments.length).toBe(3)

			expect(shape.props.segments[0].type).toBe('straight')
			expect(shape.props.segments[1].type).toBe('free')
			expect(shape.props.segments[2].type).toBe('straight')
		})

		it('Extends previously drawn line when shift is held', () => {
			editor
				.setCurrentTool(toolType)
				.keyDown('Shift')
				.pointerDown(10, 10)
				.pointerUp()
				.pointerDown(20, 20)

			const shape1 = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape1.props.segments.length).toBe(2)
			expect(shape1.props.segments[0].type).toBe('straight')
			expect(shape1.props.segments[1].type).toBe('straight')

			editor.pointerUp().pointerDown(30, 30).pointerUp()

			const shape2 = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape2.props.segments.length).toBe(3)
			expect(shape2.props.segments[2].type).toBe('straight')
		})

		it('Does not extends previously drawn line after switching to another tool', () => {
			editor
				.setCurrentTool(toolType)
				.pointerDown(10, 10)
				.pointerUp()
				.setCurrentTool('select')
				.setCurrentTool(toolType)
				.keyDown('Shift')
				.pointerDown(20, 20)
				.pointerMove(30, 30)

			expect(editor.getCurrentPageShapes()).toHaveLength(2)

			const shape1 = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape1.props.segments.length).toBe(1)
			expect(shape1.props.segments[0].type).toBe('free')

			const shape2 = editor.getCurrentPageShapes()[1] as DrawableShape
			expect(shape2.props.segments.length).toBe(1)
			expect(shape2.props.segments[0].type).toBe('straight')
		})

		it('Snaps to 15 degree angle when shift is held', () => {
			const magnitude = 10
			const angle = (17 * Math.PI) / 180
			const x = magnitude * Math.cos(angle)
			const y = magnitude * Math.sin(angle)

			const snappedAngle = (15 * Math.PI) / 180
			const snappedX = magnitude * Math.cos(snappedAngle)
			const snappedY = magnitude * Math.sin(snappedAngle)

			editor.setCurrentTool(toolType).keyDown('Shift').pointerDown(0, 0).pointerMove(x, y)

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			const segment = shape.props.segments[0]
			expect(segment.points[1].x).toBeCloseTo(snappedX)
			expect(segment.points[1].y).toBeCloseTo(snappedY)
		})

		it('Doesnt snap to 15 degree angle when cmd is held', () => {
			const magnitude = 10
			const angle = (17 * Math.PI) / 180
			const x = magnitude * Math.cos(angle)
			const y = magnitude * Math.sin(angle)

			editor.setCurrentTool(toolType).keyDown('Meta').pointerDown(0, 0).pointerMove(x, y)

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			const segment = shape.props.segments[0]
			expect(segment.points[1].x).toBeCloseTo(x)
			expect(segment.points[1].y).toBeCloseTo(y)
		})

		it('Snaps to start or end of straight segments in self when shift + cmd is held', () => {
			editor
				.setCurrentTool(toolType)
				.keyDown('Shift')
				.pointerDown(0, 0)
				.pointerUp()
				.pointerDown(0, 10)
				.pointerUp()
				.pointerDown(10, 0)
				.pointerUp()
				.pointerDown(10, 0)
				.pointerMove(1, 0) // very close to first point

			const shape1 = editor.getCurrentPageShapes()[0] as DrawableShape
			const segment1 = last(shape1.props.segments)!
			const point1 = last(segment1.points)!
			expect(point1.x).toBe(1)

			editor.keyDown('Meta')
			const shape2 = editor.getCurrentPageShapes()[0] as DrawableShape
			const segment2 = last(shape2.props.segments)!
			const point2 = last(segment2.points)!
			expect(point2.x).toBe(0)
		})

		it('Snaps to position along straight segments in self when shift + cmd is held', () => {
			editor
				.setCurrentTool(toolType)
				.keyDown('Shift')
				.pointerDown(0, 0)
				.pointerUp()
				.pointerDown(0, 10)
				.pointerUp()
				.pointerDown(10, 5)
				.pointerUp()
				.pointerDown(10, 5)
				.pointerMove(1, 5)

			const shape1 = editor.getCurrentPageShapes()[0] as DrawableShape
			const segment1 = last(shape1.props.segments)!
			const point1 = last(segment1.points)!
			expect(point1.x).toBe(1)

			editor.keyDown('Meta')
			const shape2 = editor.getCurrentPageShapes()[0] as DrawableShape
			const segment2 = last(shape2.props.segments)!
			const point2 = last(segment2.points)!
			expect(point2.x).toBe(0)
		})

		it('Deletes very short lines on interrupt', () => {
			editor.setCurrentTool(toolType).pointerDown(0, 0).pointerMove(0.1, 0.1).interrupt()
			expect(editor.getCurrentPageShapes()).toHaveLength(0)
		})

		it('Does not delete longer lines on interrupt', () => {
			editor.setCurrentTool(toolType).pointerDown(0, 0).pointerMove(5, 5).interrupt()
			expect(editor.getCurrentPageShapes()).toHaveLength(1)
		})

		it('Completes on cancel', () => {
			editor.setCurrentTool(toolType).pointerDown(0, 0).pointerMove(5, 5).cancel()
			expect(editor.getCurrentPageShapes()).toHaveLength(1)
			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.props.segments.length).toBe(1)
		})

		// New tests for pressure system
		it('Omits pressure property when no pressure is provided', () => {
			editor.setCurrentTool(toolType).pointerDown(10, 10, { pointerId: 1 }).pointerUp()

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			const point = shape.props.segments[0].points[0]
			
			// Point should not have z property when no pressure provided
			expect(point.z).toBeUndefined()
		})

		it('Stores pressure as integer 0-100 when pressure is provided', () => {
			// Simulate pen input with pressure
			editor.setCurrentTool(toolType).pointerDown(10, 10, { 
				pointerId: 1, 
				isPen: true,
				point: { x: 10, y: 10, z: 0.8 } // 80% pressure
			}).pointerUp()

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			const point = shape.props.segments[0].points[0]
			
			// Should convert 0.8 * 1.25 * 100 = 100 (clamped)
			expect(point.z).toBe(100)
			expect(Number.isInteger(point.z!)).toBe(true)
		})

		it('Handles low pressure values correctly', () => {
			// Simulate pen input with low pressure
			editor.setCurrentTool(toolType).pointerDown(10, 10, { 
				pointerId: 1, 
				isPen: true,
				point: { x: 10, y: 10, z: 0.3 } // 30% pressure
			}).pointerUp()

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			const point = shape.props.segments[0].points[0]
			
			// Should convert 0.3 * 1.25 * 100 = 37.5 -> 38 (rounded)
			expect(point.z).toBe(38)
			expect(Number.isInteger(point.z!)).toBe(true)
		})

		it('Clamps pressure values to 0-100 range', () => {
			// Simulate pen input with very high pressure that would exceed 100
			editor.setCurrentTool(toolType).pointerDown(10, 10, { 
				pointerId: 1, 
				isPen: true,
				point: { x: 10, y: 10, z: 0.9 } // 90% pressure
			}).pointerUp()

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			const point = shape.props.segments[0].points[0]
			
			// Should convert 0.9 * 1.25 * 100 = 112.5 -> 100 (clamped)
			expect(point.z).toBe(100)
			expect(Number.isInteger(point.z!)).toBe(true)
		})
	})
}

it('Draws a bunch', () => {
	editor.setCurrentTool('draw').setCamera({ x: 0, y: 0, z: 1 })

	const [first, ...rest] = TEST_DRAW_SHAPE_SCREEN_POINTS
	editor.pointerMove(first.x, first.y).pointerDown()

	for (const point of rest) {
		editor.pointerMove(point.x, point.y)
	}

	editor.pointerUp()
	editor.selectAll()

	const shape = { ...editor.getLastCreatedShape() }
	// @ts-expect-error
	delete shape.id
	expect(shape).toMatchSnapshot('draw shape')
})
