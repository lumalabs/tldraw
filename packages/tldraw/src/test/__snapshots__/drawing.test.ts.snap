// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draws a bunch: draw shape 1`] = `
{
  "index": "a1",
  "isLocked": false,
  "meta": {},
  "opacity": 1,
  "parentId": "page:page",
  "props": {
    "color": "black",
    "dash": "draw",
    "fill": "none",
    "isClosed": false,
    "isComplete": true,
    "isPen": false,
    "scale": 1,
    "segments": [
      {
        "points": [
          {
            "x": 0,
            "y": 0,
          },
          {
            "x": 1,
            "y": 0,
          },
          {
            "x": 4,
            "y": 0,
          },
          {
            "x": 10,
            "y": -1,
          },
          {
            "x": 19,
            "y": -4,
          },
          {
            "x": 30,
            "y": -10,
          },
          {
            "x": 46,
            "y": -20,
          },
          {
            "x": 61,
            "y": -30,
          },
          {
            "x": 74,
            "y": -43,
          },
          {
            "x": 89,
            "y": -59,
          },
          {
            "x": 102,
            "y": -77,
          },
          {
            "x": 108,
            "y": -90,
          },
          {
            "x": 112,
            "y": -103,
          },
          {
            "x": 117,
            "y": -119,
          },
          {
            "x": 118,
            "y": -131,
          },
          {
            "x": 119,
            "y": -137,
          },
          {
            "x": 119,
            "y": -145,
          },
          {
            "x": 120,
            "y": -152,
          },
          {
            "x": 119,
            "y": -158,
          },
          {
            "x": 117,
            "y": -163,
          },
          {
            "x": 114,
            "y": -167,
          },
          {
            "x": 109,
            "y": -169,
          },
          {
            "x": 103,
            "y": -170,
          },
          {
            "x": 97,
            "y": -170,
          },
          {
            "x": 89,
            "y": -170,
          },
          {
            "x": 80,
            "y": -166,
          },
          {
            "x": 71,
            "y": -159,
          },
          {
            "x": 62,
            "y": -150,
          },
          {
            "x": 54,
            "y": -138,
          },
          {
            "x": 50,
            "y": -126,
          },
          {
            "x": 47,
            "y": -113,
          },
          {
            "x": 46,
            "y": -99,
          },
          {
            "x": 46,
            "y": -82,
          },
          {
            "x": 47,
            "y": -61,
          },
          {
            "x": 53,
            "y": -41,
          },
          {
            "x": 60,
            "y": -24,
          },
          {
            "x": 68,
            "y": -7,
          },
          {
            "x": 79,
            "y": 12,
          },
          {
            "x": 88,
            "y": 32,
          },
          {
            "x": 96,
            "y": 50,
          },
          {
            "x": 103,
            "y": 69,
          },
          {
            "x": 106,
            "y": 86,
          },
          {
            "x": 107,
            "y": 102,
          },
          {
            "x": 107,
            "y": 120,
          },
          {
            "x": 102,
            "y": 136,
          },
          {
            "x": 90,
            "y": 146,
          },
          {
            "x": 74,
            "y": 154,
          },
          {
            "x": 43,
            "y": 163,
          },
          {
            "x": 32,
            "y": 164,
          },
          {
            "x": 21,
            "y": 164,
          },
          {
            "x": 11,
            "y": 164,
          },
          {
            "x": 2,
            "y": 164,
          },
          {
            "x": -7,
            "y": 162,
          },
          {
            "x": -13,
            "y": 159,
          },
          {
            "x": -15,
            "y": 153,
          },
          {
            "x": -15,
            "y": 147,
          },
          {
            "x": -11,
            "y": 138,
          },
          {
            "x": 1,
            "y": 127,
          },
          {
            "x": 15,
            "y": 112,
          },
          {
            "x": 34,
            "y": 96,
          },
          {
            "x": 56,
            "y": 79,
          },
          {
            "x": 81,
            "y": 58,
          },
          {
            "x": 107,
            "y": 33,
          },
          {
            "x": 126,
            "y": 12,
          },
          {
            "x": 145,
            "y": -10,
          },
          {
            "x": 160,
            "y": -30,
          },
          {
            "x": 172,
            "y": -50,
          },
          {
            "x": 185,
            "y": -73,
          },
          {
            "x": 194,
            "y": -93,
          },
          {
            "x": 199,
            "y": -112,
          },
          {
            "x": 202,
            "y": -127,
          },
          {
            "x": 203,
            "y": -138,
          },
          {
            "x": 203,
            "y": -146,
          },
          {
            "x": 201,
            "y": -152,
          },
          {
            "x": 196,
            "y": -155,
          },
          {
            "x": 191,
            "y": -156,
          },
          {
            "x": 186,
            "y": -157,
          },
          {
            "x": 178,
            "y": -156,
          },
          {
            "x": 170,
            "y": -150,
          },
          {
            "x": 164,
            "y": -140,
          },
          {
            "x": 158,
            "y": -128,
          },
          {
            "x": 151,
            "y": -110,
          },
          {
            "x": 144,
            "y": -89,
          },
          {
            "x": 139,
            "y": -64,
          },
          {
            "x": 135,
            "y": -36,
          },
          {
            "x": 132,
            "y": -7,
          },
          {
            "x": 132,
            "y": 22,
          },
          {
            "x": 132,
            "y": 49,
          },
          {
            "x": 133,
            "y": 74,
          },
          {
            "x": 140,
            "y": 97,
          },
          {
            "x": 148,
            "y": 113,
          },
          {
            "x": 156,
            "y": 124,
          },
          {
            "x": 166,
            "y": 137,
          },
          {
            "x": 175,
            "y": 145,
          },
          {
            "x": 183,
            "y": 150,
          },
          {
            "x": 191,
            "y": 152,
          },
          {
            "x": 197,
            "y": 152,
          },
          {
            "x": 205,
            "y": 151,
          },
          {
            "x": 214,
            "y": 146,
          },
          {
            "x": 223,
            "y": 136,
          },
          {
            "x": 230,
            "y": 125,
          },
          {
            "x": 236,
            "y": 112,
          },
          {
            "x": 242,
            "y": 95,
          },
          {
            "x": 247,
            "y": 78,
          },
          {
            "x": 250,
            "y": 61,
          },
          {
            "x": 252,
            "y": 46,
          },
          {
            "x": 253,
            "y": 37,
          },
          {
            "x": 253,
            "y": 31,
          },
          {
            "x": 253,
            "y": 24,
          },
          {
            "x": 251,
            "y": 20,
          },
          {
            "x": 248,
            "y": 16,
          },
          {
            "x": 246,
            "y": 16,
          },
          {
            "x": 243,
            "y": 16,
          },
          {
            "x": 240,
            "y": 17,
          },
          {
            "x": 238,
            "y": 19,
          },
          {
            "x": 236,
            "y": 26,
          },
          {
            "x": 234,
            "y": 34,
          },
          {
            "x": 233,
            "y": 45,
          },
          {
            "x": 232,
            "y": 56,
          },
          {
            "x": 232,
            "y": 66,
          },
          {
            "x": 235,
            "y": 79,
          },
          {
            "x": 241,
            "y": 91,
          },
          {
            "x": 247,
            "y": 100,
          },
          {
            "x": 255,
            "y": 109,
          },
          {
            "x": 260,
            "y": 113,
          },
          {
            "x": 266,
            "y": 116,
          },
          {
            "x": 274,
            "y": 118,
          },
          {
            "x": 280,
            "y": 118,
          },
          {
            "x": 286,
            "y": 115,
          },
          {
            "x": 291,
            "y": 105,
          },
          {
            "x": 296,
            "y": 93,
          },
          {
            "x": 298,
            "y": 83,
          },
          {
            "x": 301,
            "y": 70,
          },
          {
            "x": 303,
            "y": 58,
          },
          {
            "x": 305,
            "y": 48,
          },
          {
            "x": 306,
            "y": 38,
          },
          {
            "x": 307,
            "y": 31,
          },
          {
            "x": 308,
            "y": 25,
          },
          {
            "x": 308,
            "y": 22,
          },
          {
            "x": 308,
            "y": 20,
          },
          {
            "x": 308,
            "y": 19,
          },
          {
            "x": 308,
            "y": 22,
          },
          {
            "x": 308,
            "y": 27,
          },
          {
            "x": 308,
            "y": 35,
          },
          {
            "x": 308,
            "y": 44,
          },
          {
            "x": 308,
            "y": 51,
          },
          {
            "x": 308,
            "y": 56,
          },
          {
            "x": 308,
            "y": 61,
          },
          {
            "x": 309,
            "y": 66,
          },
          {
            "x": 312,
            "y": 71,
          },
          {
            "x": 314,
            "y": 74,
          },
          {
            "x": 317,
            "y": 75,
          },
          {
            "x": 320,
            "y": 76,
          },
          {
            "x": 324,
            "y": 76,
          },
          {
            "x": 329,
            "y": 73,
          },
          {
            "x": 333,
            "y": 69,
          },
          {
            "x": 336,
            "y": 66,
          },
          {
            "x": 339,
            "y": 62,
          },
          {
            "x": 342,
            "y": 59,
          },
          {
            "x": 344,
            "y": 57,
          },
          {
            "x": 346,
            "y": 55,
          },
          {
            "x": 348,
            "y": 55,
          },
          {
            "x": 348,
            "y": 55,
          },
          {
            "x": 349,
            "y": 55,
          },
          {
            "x": 349,
            "y": 56,
          },
          {
            "x": 350,
            "y": 57,
          },
          {
            "x": 351,
            "y": 59,
          },
          {
            "x": 351,
            "y": 61,
          },
          {
            "x": 352,
            "y": 62,
          },
          {
            "x": 352,
            "y": 63,
          },
          {
            "x": 353,
            "y": 64,
          },
          {
            "x": 354,
            "y": 64,
          },
          {
            "x": 355,
            "y": 64,
          },
          {
            "x": 356,
            "y": 58,
          },
          {
            "x": 358,
            "y": 49,
          },
          {
            "x": 360,
            "y": 40,
          },
          {
            "x": 363,
            "y": 32,
          },
          {
            "x": 365,
            "y": 26,
          },
          {
            "x": 367,
            "y": 19,
          },
          {
            "x": 369,
            "y": 13,
          },
          {
            "x": 373,
            "y": 7,
          },
          {
            "x": 376,
            "y": 3,
          },
          {
            "x": 380,
            "y": 2,
          },
          {
            "x": 385,
            "y": 2,
          },
          {
            "x": 390,
            "y": 2,
          },
          {
            "x": 397,
            "y": 3,
          },
          {
            "x": 410,
            "y": 11,
          },
          {
            "x": 424,
            "y": 23,
          },
          {
            "x": 434,
            "y": 34,
          },
          {
            "x": 446,
            "y": 49,
          },
          {
            "x": 456,
            "y": 64,
          },
          {
            "x": 464,
            "y": 81,
          },
          {
            "x": 468,
            "y": 95,
          },
          {
            "x": 470,
            "y": 116,
          },
          {
            "x": 472,
            "y": 142,
          },
          {
            "x": 472,
            "y": 162,
          },
          {
            "x": 468,
            "y": 178,
          },
          {
            "x": 458,
            "y": 195,
          },
          {
            "x": 442,
            "y": 213,
          },
          {
            "x": 423,
            "y": 230,
          },
          {
            "x": 407,
            "y": 240,
          },
          {
            "x": 393,
            "y": 245,
          },
          {
            "x": 377,
            "y": 250,
          },
          {
            "x": 364,
            "y": 252,
          },
          {
            "x": 354,
            "y": 252,
          },
          {
            "x": 346,
            "y": 248,
          },
          {
            "x": 340,
            "y": 239,
          },
          {
            "x": 339,
            "y": 225,
          },
          {
            "x": 339,
            "y": 198,
          },
          {
            "x": 349,
            "y": 165,
          },
          {
            "x": 372,
            "y": 130,
          },
          {
            "x": 403,
            "y": 89,
          },
          {
            "x": 432,
            "y": 54,
          },
          {
            "x": 467,
            "y": 16,
          },
          {
            "x": 504,
            "y": -21,
          },
          {
            "x": 551,
            "y": -68,
          },
          {
            "x": 597,
            "y": -115,
          },
          {
            "x": 619,
            "y": -138,
          },
          {
            "x": 641,
            "y": -162,
          },
          {
            "x": 663,
            "y": -188,
          },
          {
            "x": 675,
            "y": -203,
          },
          {
            "x": 684,
            "y": -219,
          },
          {
            "x": 692,
            "y": -237,
          },
          {
            "x": 693,
            "y": -244,
          },
          {
            "x": 691,
            "y": -250,
          },
          {
            "x": 682,
            "y": -254,
          },
          {
            "x": 664,
            "y": -256,
          },
          {
            "x": 642,
            "y": -256,
          },
          {
            "x": 621,
            "y": -253,
          },
          {
            "x": 589,
            "y": -240,
          },
          {
            "x": 554,
            "y": -221,
          },
          {
            "x": 526,
            "y": -201,
          },
          {
            "x": 502,
            "y": -182,
          },
          {
            "x": 484,
            "y": -165,
          },
          {
            "x": 467,
            "y": -146,
          },
          {
            "x": 456,
            "y": -131,
          },
          {
            "x": 450,
            "y": -120,
          },
          {
            "x": 448,
            "y": -112,
          },
          {
            "x": 448,
            "y": -107,
          },
          {
            "x": 449,
            "y": -104,
          },
          {
            "x": 452,
            "y": -103,
          },
          {
            "x": 458,
            "y": -102,
          },
          {
            "x": 462,
            "y": -102,
          },
          {
            "x": 465,
            "y": -103,
          },
          {
            "x": 470,
            "y": -104,
          },
          {
            "x": 472,
            "y": -105,
          },
          {
            "x": 474,
            "y": -106,
          },
          {
            "x": 475,
            "y": -106,
          },
          {
            "x": 476,
            "y": -107,
          },
          {
            "x": 476,
            "y": -107,
          },
          {
            "x": 477,
            "y": -107,
          },
        ],
        "type": "free",
      },
    ],
    "size": "m",
  },
  "rotation": 0,
  "type": "draw",
  "typeName": "shape",
  "x": 511,
  "y": 234,
}
`;
